PATCH-P0-FIX-UPSTREAM: Generate shared libmysqld

In Linux distributions, we love shared libraries and we want to link gainst
shared libraries. As libmysqld library is used for example by Amarok, we want
this library to be shared.

Maintainer: Michal Hrusecky <Michal.Hrusecky@opensuse.org>

Index: libmysqld/CMakeLists.txt
===================================================================
--- libmysqld/CMakeLists.txt.orig
+++ libmysqld/CMakeLists.txt
@@ -138,7 +138,17 @@ IF(UNIX)
   ${CMAKE_STATIC_LIBRARY_PREFIX}mysqld-debug)
 ENDIF()
 
-IF(MSVC AND NOT DISABLE_SHARED)
-  MERGE_LIBRARIES(libmysqld SHARED ${LIBS} EXPORTS ${CLIENT_API_FUNCTIONS}
-  COMPONENT Embedded)
+IF(NOT DISABLE_SHARED)
+  MERGE_LIBRARIES(libmysqld SHARED mysqlserver EXPORTS ${CLIENT_API_FUNCTIONS})
+  IF(UNIX)
+    # Name the shared library, handle versioning (provides same api as client library
+    # hence the same version)
+    SET_TARGET_PROPERTIES(libmysqld PROPERTIES
+      OUTPUT_NAME mysqld
+      VERSION "${SHARED_LIB_MAJOR_VERSION}.0.0"
+      SOVERSION "${SHARED_LIB_MAJOR_VERSION}")
+    # Clean direct output flags, as 2 targets have the same base name (libmysqld)
+    SET_TARGET_PROPERTIES(libmysqld PROPERTIES CLEAN_DIRECT_OUTPUT 1)
+    SET_TARGET_PROPERTIES(mysqlserver PROPERTIES CLEAN_DIRECT_OUTPUT 1)
+  ENDIF()
 ENDIF()
Index: cmake/libutils.cmake
===================================================================
--- cmake/libutils.cmake.orig
+++ cmake/libutils.cmake
@@ -268,6 +268,16 @@ MACRO(MERGE_LIBRARIES)
     MYSQL_INSTALL_TARGETS(${TARGET} DESTINATION "${INSTALL_LIBDIR}" ${COMP})
   ENDIF()
   SET_TARGET_PROPERTIES(${TARGET} PROPERTIES LINK_INTERFACE_LIBRARIES "")
+
+  IF(ARG_SHARED AND LINK_FLAG_NO_UNDEFINED)
+    # Do not allow undefined symbols in shared libraries
+    GET_TARGET_PROPERTY(TARGET_LINK_FLAGS ${TARGET} LINK_FLAGS)
+    IF(NOT TARGET_LINK_FLAGS)
+      SET(TARGET_LINK_FLAGS)
+    ENDIF()
+    SET_TARGET_PROPERTIES(${TARGET} PROPERTIES LINK_FLAGS
+      "${TARGET_LINK_FLAGS} ${LINK_FLAG_NO_UNDEFINED}")
+  ENDIF()
 ENDMACRO()
 
 FUNCTION(GET_DEPENDEND_OS_LIBS target result)
